### PR TITLE
Add Iterator supports for v-for

### PR DESCRIPTION
**Pull request by lyh2668**
添加v-for循环对迭代器(Iterator)的支持

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: src/core/instance/render-helpers/render-list.js [hunk 1]
```diff
@@ -26,10 +26,25 @@ export function renderList (
     }
   } else if (isObject(val)) {
     keys = Object.keys(val)
-    ret = new Array(keys.length)
-    for (i = 0, l = keys.length; i < l; i++) {
-      key = keys[i]
-      ret[i] = render(val[key], key, i)
+    if (keys.length === 0 && val.toString().indexOf('Iterator') > -1) {
```
Comment (HerringtonDarkholme) [inline src/core/instance/render-helpers/render-list.js]: `Iterator` looks like new to me. Would you like to give me a link of introduction to `Iterator`? I might be wrong but Google only gave me `Generator`.

Comment (lyh2668) [reply src/core/instance/render-helpers/render-list.js]: 1.  [Map.prototype.keys()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/keys) =>return a new Map iterator object.   
2. [Set.prototype.values()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/values) => return a new Iterator object containing the values for each element in the given Set, in insertion order.

Comment (lyh2668) [reply src/core/instance/render-helpers/render-list.js]: [Iteration_protocols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols)

Comment (HerringtonDarkholme) [reply src/core/instance/render-helpers/render-list.js]: Thanks. I have checked the spec and MapIterator's `@@toStringTag` symbol returns `Map Iterator`. The similar for other iterators.

GIven this, I suspect if a `toString` test is legitimate. Checking `Symbol.iterator` in proper environment is better, I think.

Comment (lyh2668) [reply src/core/instance/render-helpers/render-list.js]: Yes, the better way is checking Symbol.iterator. At first I thought I can't use the ES6 grammar.

Comment (lyh2668) [reply src/core/instance/render-helpers/render-list.js]: Symbol.toStringTag
A string value used for the default description of an object. Used by Object.prototype.toString().
[Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)

Comment (lyh2668) [reply src/core/instance/render-helpers/render-list.js]: Symbol.toStringTag
A string value used for the default description of an object. Used by Object.prototype.toString().
[Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)

Comment (lyh2668) [reply src/core/instance/render-helpers/render-list.js]: `Symbol.toStringTag`
A string value used for the default description of an object. Used by `Object.prototype.toString()`.
[Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)

**Comment by Gudradain** [comment]
Iterator can be infinite no? Would that be a problem?

**Comment by blai** [comment]
echo @Gudradain's comment, we need to add a limit to this (maybe with a default value), but the same could go for traditional Array with excessive amount of data (so it's virtually infinite)

**Comment by mitar** [comment]
Related: #5893

**Comment by mitar** [comment]
So I have been using the following for a while now: https://github.com/vuejs/vue/blob/cb062c50cdd21628dcf758fe54c84d17d75cb2ca/src/core/instance/render-helpers/render-list.js

But I thought that in #5893 it is said that this will not go into core for now?

**Comment by mitar** [comment]
I think #8179 is a better implementation of this. Better check for iterator support.

**Comment by yyx990803** [comment]
Thanks for the PR, closing in favor of #8179 due to better support check and tests.

**Pull request closed without merging**